modules/gmysqlbackend/smysql.cc

```diff
@@ -19,41 +19,27 @@ pthread_mutex_t SMySQL::s_myinitlock = PTHREAD_MUTEX_INITIALIZER;
 class SMySQLStatement: public SSqlStatement
 {
 public:
-  SMySQLStatement(const string& query, bool dolog, int nparams, MYSQL* db) 
+  SMySQLStatement(const string& query, bool dolog, int nparams, MYSQL* db) : d_prepared(false)
   {
-    int err;
     d_db = db;
     d_dolog = dolog;
     d_query = query;
-    d_parnum = d_paridx = d_fnum = d_resnum = d_residx = 0;
+    d_paridx = d_fnum = d_resnum = d_residx = 0;
+    d_parnum = nparams;
     d_req_bind = d_res_bind = NULL;
     d_stmt = NULL;
 
     if (query.empty()) {
       return;
     }
-
-    if ((d_stmt = mysql_stmt_init(d_db))==NULL) 
-      throw SSqlException("Could not initialize mysql statement, out of memory: " + d_query);
-    
-    if ((err = mysql_stmt_prepare(d_stmt, query.c_str(), query.size()))) {
-      string error(mysql_stmt_error(d_stmt));
-      throw SSqlException("Could not prepare statement: " + d_query + string(": ") + error);
-    }
-
-    if (static_cast<int>(mysql_stmt_param_count(d_stmt)) != nparams) 
-      throw SSqlException("Provided parameter count does not match statement: " + d_query);
-   
-    d_parnum = nparams;
-    if (d_parnum>0) {
-      d_req_bind = new MYSQL_BIND[d_parnum];
-      memset(d_req_bind, 0, sizeof(MYSQL_BIND)*d_parnum);
-    }
   }
 
   SSqlStatement* bind(const string& name, bool value) {
-    if (d_paridx >= d_parnum) 
+    prepareStatement();
+    if (d_paridx >= d_parnum) {
+      releaseStatement();
```

Comment:
we are not going to destroy the statement in the future if I've understood correctly, but yes, in this case it probably should be and then dtor would be handling it.

I would welcome any input on how to do signalling about when reprepare won't save you and when it does.